lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm

```diff
@@ -709,4 +770,54 @@ sub scalar_property_from_underlying_alignment_results {
 }
 
 
+sub _lock_per_lane_alignment {
+    my $self = shift;
+
+    for my $alignment ($self->collect_individual_alignments) {
+        unless ($alignment->get_merged_alignment_results) {
+            my @bams = $alignment->alignment_bam_file_paths;
+            unless (@bams) {
+               die $self->error_message("Alignment result with class (%s) and id (%s) has neither ".
+                   "merged results nor valid bam paths. This likely means that this alignment result ".
+                   "needs to be removed and realigned because data has been lost. ".
+                   "Please create an apipe-support ticket for this.", $alignment->class, $alignment->id);
+               #There is no way to recreate per lane bam if merged bam
+               #does not exist and per lane bam is removed. Software
+               #result of this per lane alignment needs to be removed and
+               #this per lane instrument data needs to be realigned
+               #with that aligner.
+            }
+
+            unless ($ENV{UR_DBI_NO_COMMIT}) {
+                my $lock_var = File::Spec->join($ENV{GENOME_LOCK_DIR}, 'genome', __PACKAGE__, 'lock-per-lane-alignment-'.$alignment->id);
+                my $lock = Genome::Sys->lock_resource(
+                    resource_lock => $lock_var,
+                    scope         => 'site',
+                    max_try       => 288, # Try for 48 hours every 10 minutes
+                    block_sleep   => 600,
+                );
+                die $self->error_message("Unable to acquire the lock for per lane alignment result id (%s) !", $alignment->id) unless $lock;
+
+                # If the build before us successfully created a merged alignment result, we no longer need a lock
+                # If it failed, we will add an observer just as the first build did.
+                if ($alignment->get_merged_alignment_results) {
```

Comment:
Hmm. Good point. I'd have to think about this for little bit. I think I like your suggestion.

If we went forward with that, we'd still check again and (possibly) immediately unlock, correct? I believe that is the case, in the case of contention around the lock, and 'waking up' to find the merged alignment result is now present. So I believe the only code change would be to add one additional check before locking and skip locking if appropriate?